File: ios/BiblosSurfer/Services/Services/LibraryService.swift
//
//  LibraryService.swift
//  BiblosSurfer
//
//  Created by Mikołaj Linczewski on 21/08/2026.
//

import Foundation
import UIKit

protocol LibraryServiceProtocol {
    func loadItems() async throws -> [LibraryItem]
    func importBook(from sourceURL: URL) async throws -> LibraryItem
}

final class LibraryService: LibraryServiceProtocol {
    private let fileManager: FileManager
    private let booksDirectory: URL
    private let coversDirectory: URL
    private let bundledBookURL: URL?
    private let opener: PublicationOpeningServiceProtocol
    private let bookStore: BookStoreProtocol

    init(
        fileManager: FileManager = .default,
        booksDirectory: URL? = nil,
        coversDirectory: URL? = nil,
        bundledBookURL: URL? = Bundle.main.url(forResource: "genesis-ksiega-rodzaju-bereszit", withExtension: "epub"),
        opener: PublicationOpeningServiceProtocol = PublicationOpeningService(),
        bookStore: BookStoreProtocol = LocalBookService()
    ) {
        self.fileManager = fileManager
        self.booksDirectory = booksDirectory ?? fileManager.defaultBooksDirectory
        self.coversDirectory = coversDirectory ?? fileManager.defaultCoversDirectory
        self.bundledBookURL = bundledBookURL
        self.opener = opener
        self.bookStore = bookStore
    }

    func loadItems() async throws -> [LibraryItem] {
        try ensureDirectories()
        let urls = try libraryFileURLs()
        for url in urls where bookStore.book(atRelativePath: url.lastPathComponent) == nil {
            _ = try? await catalog(fileURL: url)
        }
        return urls
            .map { url in
                if let stored = bookStore.book(atRelativePath: url.lastPathComponent) {
                    return stored.replacingFileURL(url)
                }
                return LibraryItem(fileURL: url)
            }
            .sorted { $0.title.localizedCaseInsensitiveCompare($1.title) == .orderedAscending }
    }

    func importBook(from sourceURL: URL) async throws -> LibraryItem {
        let accessing = sourceURL.startAccessingSecurityScopedResource()
        defer {
            if accessing {
                sourceURL.stopAccessingSecurityScopedResource()
            }
        }

        let fileExtension = sourceURL.pathExtension.lowercased()
        guard fileExtension == "epub" || fileExtension == "pdf" else {
            throw Errors.Library.unsupportedFormat(fileExtension: fileExtension)
        }

        try ensureDirectories()
        let destination = uniqueDestination(for: sourceURL.lastPathComponent)
        do {
            try fileManager.copyItem(at: sourceURL, to: destination)
        } catch {
            throw Errors.Library.copyFailed(
                fileName: sourceURL.lastPathComponent,
                underlying: error.localizedDescription
            )
        }
        return try await catalog(fileURL: destination)
    }

    private func libraryFileURLs() throws -> [URL] {
        var urls = try listedBookURLs()
        if let bundledBookURL,
           !urls.contains(where: { $0.lastPathComponent == bundledBookURL.lastPathComponent }) {
            urls.append(bundledBookURL)
        }
        return urls
    }

    private func catalog(fileURL: URL) async throws -> LibraryItem {
        let opened: OpenedPublication
        do {
            opened = try await opener.open(url: fileURL)
        } catch let error as DescriptiveError {
            throw error
        } catch {
            throw Errors.Publication.openFailed(
                title: fileURL.deletingPathExtension().lastPathComponent,
                underlying: error.localizedDescription
            )
        }

        var coverFileName: String?
        if let cover = opened.cover, let data = cover.jpegData(compressionQuality: 0.8) {
            let name = fileURL.deletingPathExtension().lastPathComponent + ".jpg"
            let coverURL = coversDirectory.appendingPathComponent(name)
            try? data.write(to: coverURL, options: .atomic)
            coverFileName = name
        }

        let relativePath = fileURL.lastPathComponent
        let existing = bookStore.book(atRelativePath: relativePath)
        let item = LibraryItem(
            id: existing?.id ?? UUID(),
            fileName: fileURL.lastPathComponent,
            title: opened.title,
            author: opened.author,
            fileURL: fileURL,
            coverURL: coverFileName.map { coversDirectory.appendingPathComponent($0) },
            locatorJSON: existing?.locatorJSON,
            progression: existing?.progression ?? 0,
            format: opened.format,
            folderName: existing?.folderName
        )
        bookStore.upsert(item, relativePath: relativePath, coverFileName: coverFileName)
        return item
    }

    private func ensureDirectories() throws {
        do {
            try fileManager.createDirectory(at: booksDirectory, withIntermediateDirectories: true)
            try fileManager.createDirectory(at: coversDirectory, withIntermediateDirectories: true)
        } catch {
            throw Errors.Library.copyFailed(
                fileName: booksDirectory.lastPathComponent,
                underlying: error.localizedDescription
            )
        }
    }

    private func listedBookURLs() throws -> [URL] {
        let contents: [URL]
        do {
            contents = try fileManager.contentsOfDirectory(
                at: booksDirectory,
                includingPropertiesForKeys: nil,
                options: [.skipsHiddenFiles]
            )
        } catch {
            throw Errors.Library.copyFailed(
                fileName: booksDirectory.lastPathComponent,
                underlying: error.localizedDescription
            )
        }
        return contents.filter {
            let ext = $0.pathExtension.lowercased()
            return ext == "epub" || ext == "pdf"
        }
    }

    private func uniqueDestination(for fileName: String) -> URL {
        var destination = booksDirectory.appendingPathComponent(fileName)
        let baseName = destination.deletingPathExtension().lastPathComponent
        let fileExtension = destination.pathExtension
        var suffix = 1
        while fileManager.fileExists(atPath: destination.path) {
            destination = booksDirectory.appendingPathComponent("\(baseName) \(suffix).\(fileExtension)")
            suffix += 1
        }
        return destination
    }
}

private extension LibraryItem {
    func replacingFileURL(_ fileURL: URL) -> LibraryItem {
        LibraryItem(
            id: id,
            fileName: fileName,
            title: title,
            author: author,
            fileURL: fileURL,
            coverURL: coverURL,
            locatorJSON: locatorJSON,
            progression: progression,
            format: format,
            folderName: folderName
        )
    }
}

extension FileManager {
    var defaultBooksDirectory: URL {
        urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("Books", isDirectory: true)
    }

    var defaultCoversDirectory: URL {
        urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("Covers", isDirectory: true)
    }
}

File: android/app/src/main/java/miko/biblossurfer/data/LibraryService.kt
package miko.biblossurfer.data

import android.content.Context
import android.graphics.Bitmap
import miko.biblossurfer.data.model.LibraryItem
import miko.biblossurfer.data.model.PublicationFormat
import java.io.File
import java.io.FileOutputStream
import java.util.UUID

interface LibraryServiceProtocol {
    suspend fun loadItems(): List<LibraryItem>
    suspend fun importBook(from: File): LibraryItem
}

class LibraryService(
    private val booksDirectory: File,
    private val coversDirectory: File,
    private val bundledBookFiles: List<File>,
    private val opener: PublicationOpeningServiceProtocol,
    private val bookStore: BookStoreProtocol,
) : LibraryServiceProtocol {

    constructor(
        context: Context,
        opener: PublicationOpeningServiceProtocol = PublicationOpeningService(context),
        bookStore: BookStoreProtocol = LocalBookService(),
        bundledBookFiles: List<File> = context.bundledEpubFiles(),
    ) : this(
        booksDirectory = context.defaultBooksDirectory(),
        coversDirectory = context.defaultCoversDirectory(),
        bundledBookFiles = bundledBookFiles,
        opener = opener,
        bookStore = bookStore,
    )

    override suspend fun loadItems(): List<LibraryItem> {
        ensureDirectories()
        val urls = libraryFileURLs()
        for (url in urls) {
            if (bookStore.book(url.name) == null) {
                runCatching { catalog(url) }
            }
        }
        return urls
            .map { url ->
                bookStore.book(url.name)?.replacingFileURL(url) ?: LibraryItem(fileURL = url)
            }
            .sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.title })
    }

    override suspend fun importBook(from: File): LibraryItem {
        val fileExtension = from.extension.lowercase()
        if (fileExtension != "epub" && fileExtension != "pdf") {
            throw Errors.Library.UnsupportedFormat(fileExtension)
        }
        ensureDirectories()
        val destination = uniqueDestination(from.name)
        try {
            from.copyTo(destination, overwrite = false)
        } catch (error: Exception) {
            throw Errors.Library.CopyFailed(from.name, error.localizedMessage ?: error.toString())
        }
        return catalog(destination)
    }

    private fun libraryFileURLs(): List<File> {
        val urls = listedBookURLs().toMutableList()
        val existingNames = urls.map { it.name }.toSet()
        for (bundled in bundledBookFiles) {
            if (bundled.name !in existingNames) {
                urls.add(bundled)
            }
        }
        return urls
    }

    private suspend fun catalog(fileURL: File): LibraryItem {
        val opened = try {
            opener.open(fileURL)
        } catch (error: Exception) {
            if (error is DescriptiveError) throw error
            throw Errors.Publication.OpenFailed(
                fileURL.nameWithoutExtension,
                error.localizedMessage ?: error.toString(),
            )
        }

        var coverFileName: String? = null
        opened.cover?.let { cover ->
            val name = fileURL.nameWithoutExtension + ".jpg"
            val coverURL = File(coversDirectory, name)
            runCatching {
                FileOutputStream(coverURL).use { out ->
                    cover.compress(Bitmap.CompressFormat.JPEG, 80, out)
                }
                coverFileName = name
            }
        }

        val relativePath = fileURL.name
        val existing = bookStore.book(relativePath)
        val item = LibraryItem(
            id = existing?.id ?: UUID.randomUUID(),
            fileName = fileURL.name,
            title = opened.title,
            author = opened.author,
            fileURL = fileURL,
            coverURL = coverFileName?.let { File(coversDirectory, it) },
            locatorJSON = existing?.locatorJSON,
            progression = existing?.progression ?: 0.0,
            format = opened.format,
            folderName = existing?.folderName,
        )
        bookStore.upsert(item, relativePath, coverFileName)
        return item
    }

    private fun ensureDirectories() {
        try {
            booksDirectory.mkdirs()
            coversDirectory.mkdirs()
        } catch (error: Exception) {
            throw Errors.Library.CopyFailed(
                booksDirectory.name,
                error.localizedMessage ?: error.toString(),
            )
        }
    }

    private fun listedBookURLs(): List<File> {
        val contents = try {
            booksDirectory.listFiles() ?: emptyArray()
        } catch (error: Exception) {
            throw Errors.Library.CopyFailed(
                booksDirectory.name,
                error.localizedMessage ?: error.toString(),
            )
        }
        return contents.filter { file ->
            val ext = file.extension.lowercase()
            !file.isHidden && (ext == "epub" || ext == "pdf")
        }
    }

    private fun uniqueDestination(fileName: String): File {
        var destination = File(booksDirectory, fileName)
        val baseName = destination.nameWithoutExtension
        val fileExtension = destination.extension
        var suffix = 1
        while (destination.exists()) {
            destination = File(booksDirectory, "$baseName $suffix.$fileExtension")
            suffix += 1
        }
        return destination
    }
}

fun Context.defaultBooksDirectory(): File = File(filesDir, "Books")

fun Context.defaultCoversDirectory(): File = File(filesDir, "Covers")

/**
 * Bundled EPUBs live in `assets/books/`. Android assets are not real files, so they are extracted
 * once into `filesDir/BundledBooks` — not into `Books/`, matching iOS (bundle URLs stay outside
 * the user library directory).
 */
fun Context.bundledEpubFiles(): List<File> {
    val destDir = File(filesDir, "BundledBooks")
    destDir.mkdirs()
    val names = assets.list("books").orEmpty().filter { it.endsWith(".epub", ignoreCase = true) }
    return names.map { name ->
        val dest = File(destDir, name)
        if (!dest.exists()) {
            assets.open("books/$name").use { input ->
                dest.outputStream().use { output -> input.copyTo(output) }
            }
        }
        dest
    }
}

fun Context.bundledPdfFile(): File? {
    val destDir = File(filesDir, "BundledBooks")
    destDir.mkdirs()
    val name = assets.list("books").orEmpty().firstOrNull { it.endsWith(".pdf", ignoreCase = true) }
        ?: return null
    val dest = File(destDir, name)
    if (!dest.exists()) {
        assets.open("books/$name").use { input ->
            dest.outputStream().use { output -> input.copyTo(output) }
        }
    }
    return dest
}
